server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java

```diff
@@ -85,18 +145,70 @@ public String[] excludes() {
         return this.excludes;
     }
 
+    public Boolean includeVectors() {
+        return this.includeVectors;
+    }
+
+    public boolean filterVectorFields() {
+        return this.includeVectors == null || this.includeVectors == false;
+    }
+
     public boolean hasFilter() {
         return this.includes.length > 0 || this.excludes.length > 0;
     }
 
     public SourceFilter filter() {
-        return new SourceFilter(includes, excludes);
+        return filter(null);
+    }
+
+    public SourceFilter filter(MappingLookup mappingLookup) {
+        if (filterVectorFields() == false) {
+            return new SourceFilter(includes, excludes);
+        } else {
+            if (mappingLookup == null) {
+                throw new IllegalStateException("mappingLookup must not be null when filtering vectors");
+            }
+
+            String[] excludeFields = excludes();
+            List<String> inferenceFieldNames = new ArrayList<>(mappingLookup.inferenceFields().size());
+            for (String inferenceField : mappingLookup.inferenceFields().keySet()) {
+                Mapper mapper = mappingLookup.getMapper(inferenceField);
+                if (mapper instanceof InferenceFieldMapper == false) {
+                    throw new IllegalStateException("Mapper for field [" + inferenceField + "] is not an inference field mapper");
+                }
+
+                InferenceFieldMapper inferenceFieldMapper = (InferenceFieldMapper) mapper;
+                inferenceFieldNames.add(inferenceFieldMapper.getInferenceFieldName());
+            }
+            excludeFields = ArrayUtils.concat(excludeFields, inferenceFieldNames.toArray(new String[0]));
+
+            if (includeVectors != null) {
+                String[] denseVectors = mappingLookup.getFullNameToFieldType()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() instanceof DenseVectorFieldMapper.DenseVectorFieldType)
+                    .map(Map.Entry::getKey)
+                    .toArray(String[]::new);
+                excludeFields = ArrayUtils.concat(excludeFields, denseVectors);
+
+                String[] sparseVectors = mappingLookup.getFullNameToFieldType()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() instanceof SparseVectorFieldMapper.SparseVectorFieldType)
+                    .map(Map.Entry::getKey)
+                    .toArray(String[]::new);
+                excludeFields = ArrayUtils.concat(excludeFields, sparseVectors);
+            }
+
+            return new SourceFilter(includes, excludeFields);
```

Comment:
See my comments above, we should return `null` when both includes and excludeFields remain `null` (e.g.: when there is no vector field on the mapping and includeVectors is false).